src/site/content/en/fast/debugging-layout-shifts/index.md

```diff
@@ -0,0 +1,457 @@
+---
+layout: post
+title: Debugging layout shifts
+authors:
+  - katiehempenius
+date: 2021-03-11
+hero: image/j2RDdG43oidUy6AL6LovThjeX9c2/N65l8ccOUEDESpnnEkje.png
+alt: Picture of a layout shift.
+description: |
+  Learn how to identify and fix layout shifts.
+tags:
+  - blog
+  - performance
+---
+
+
+The first part of this article discusses tooling for debugging layout shifts;
+the second part of this article discusses the thought process to use when
+identifying the cause of a layout shift.
+
+
+## Tooling
+
+
+### Layout Instability API
+
+The [Layout Instability API](https://wicg.github.io/layout-instability/) is the
+browser mechanism for measuring and reporting layout shifts. All tools for
+debugging layout shifts, including DevTools, are ultimately built upon the
+Layout Instability API. However, using the Layout Instability API directly is a
+powerful debugging tool due to its flexibility.
+
+{% Aside %}
+The Layout Instability API is only
+[supported](https://caniuse.com/mdn-api_layoutshift) by Chromium browsers. At
+the current time there is no way to measure or debug layout shifts in
+non-Chromium browsers.
+{% endAside %}
+
+
+#### Usage
+
+The same code [snippet](https://web.dev/cls/#measure-cls-in-javascript) that is
+used to measure [cumulative layout shift (CLS)](https://web.dev/cls/) can also
+be used to debug layout shifts. The snippet below logs information about layout
+shifts to the console. Inspecting this information will provide you information
+about when, where, and how a layout shift occurred.
+
+```javascript
+let cls = 0;
+new PerformanceObserver((entryList) => {
+  for (const entry of entryList.getEntries()) {
+    if (!entry.hadRecentInput) {
+      cls += entry.value;
+      console.log('Current CLS value:', cls, entry);
+    }
+  }
+}).observe({type: 'layout-shift', buffered: true});
+```
+
+When running this script be aware that:
+
+
+
+*   The `buffered: true` option indicates that the
+    [PerformanceObserver](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceObserver)
+    should check the browser's [performance entry
+    buffer](https://www.w3.org/TR/performance-timeline-2/#dfn-performance-entry-buffer)
+    for performance entries that were created before the observer's
+    initialization. As a result, the PerformanceObserver will report layout
+    shifts that happened both before and after it was initialized. Keep this in
+    mind when inspecting the console logs. An initial glut of layout shifts can
+    reflect a reporting backlog, rather than the sudden occurrence of numerous
+    layout shifts.
```

Comment:
```suggestion
*   The `buffered: true` option indicates that the
    [`PerformanceObserver`](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceObserver)
    should check the browser's [performance entry
    buffer](https://www.w3.org/TR/performance-timeline-2/#dfn-performance-entry-buffer)
    for performance entries that were created before the observer's
    initialization. As a result, the `PerformanceObserver` will report layout
    shifts that happened both before and after it was initialized. Keep this in
    mind when inspecting the console logs. An initial glut of layout shifts can
    reflect a reporting backlog, rather than the sudden occurrence of numerous
    layout shifts.
```